include/onnxruntime/core/session/onnxruntime_c_api.h

```diff
@@ -545,10 +545,10 @@ struct OrtApi {
   // Always returns the same instance on every invocation.
   OrtStatus*(ORT_API_CALL* GetAllocatorWithDefaultOptions)(_Outptr_ OrtAllocator** out)NO_EXCEPTION;
 
-  // Override symbolic dimensions with actual values if known at session initialization time to enable
+  // Override symbolic dimensions (by specific denotation strings) with actual values if known at session initialization time to enable
   // optimizations that can take advantage of fixed values (such as memory planning, etc)
-  OrtStatus*(ORT_API_CALL* AddFreeDimensionOverride)(_Inout_ OrtSessionOptions* options,
```

Comment:
Unfortunately removing this will break binary compatibility. We can only mark existing APIs as deprecated; we can't remove them. I suggest we keep this API and document it saying it's based on denotation. The other option is to mark this API as deprecated and add the 2 new APIs.